pkg/apis/ceph.rook.io/v1/validate_resources_test.go

```diff
@@ -0,0 +1,120 @@
+package v1
+
+import (
+	rookv1 "github.com/rook/rook/pkg/apis/rook.io/v1"
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"testing"
+)
+
+func TestCephClusterValidateCreate(t *testing.T) {
+	c := getCephClusterCRD()
+	err := c.ValidateCreate()
+	assert.NoError(t, err)
+	c.Spec.External.Enable = true
+	c.Spec.Monitoring = MonitoringSpec{}
+	err = c.ValidateCreate()
+	assert.Error(t, err)
+}
+
+func TestValidatePoolSpec(t *testing.T) {
+	p := getCephBlockPoolCRD()
+	err := ValidatePoolSpecs(p.Spec)
+	assert.NoError(t, err)
+
+	p.Spec.ErasureCoded.DataChunks = 1
+	err = ValidatePoolSpecs(p.Spec)
+	assert.Error(t, err)
+}
+
+func TestCephBlockPoolValidateUpdate(t *testing.T) {
+	p := getCephBlockPoolCRD()
+	up := p.DeepCopy()
+	up.Spec.Replicated.Size = 3
+	up.Spec.Replicated.RequireSafeReplicaSize = true
+	err := up.ValidateUpdate(p)
+	assert.Error(t, err)
+}
+
+func TestCephClusterValidateUpdate(t *testing.T) {
+	c := getCephClusterCRD()
+	err := c.ValidateCreate()
+	assert.NoError(t, err)
+
+	// Updating the CRD specs with invalid values
+	uc := c.DeepCopy()
+	uc.Spec.DataDirHostPath = "var/rook"
+	err = uc.ValidateUpdate(c)
+	assert.Error(t, err)
+}
+
+func getCephBlockPoolCRD() *CephBlockPool {
+	return &CephBlockPool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "ec-pool",
+			Namespace: "rook-ceph",
+		},
+		Spec: PoolSpec{
+			FailureDomain: "osd",
+			ErasureCoded: ErasureCodedSpec{
+				CodingChunks: 1,
+				DataChunks:   2,
+			},
+			Parameters: map[string]string{
+				"compression_mode": "none",
+			},
+		},
+	}
+}
+func getCephClusterCRD() *CephCluster {
+	return &CephCluster{
```

Comment:
Much better instead of yaml, thanks. One more point is that there is no need for the test to set all the fields. Just set what is needed for an individual test instead of all the optional fields. With that approach, you likely don't even need a helper to create the struct and you can just create a small struct in each individual test.

Reply:
Well i was thinking about not adding any duplicates in code but I've removed the helper funcs.